### DISCUSSION TITLE
Sensitivities for an experiment with thermal considerations

### DISCUSSION THREAD

**Discussion by alexgUCSD** [pybamm-team/PyBaMM#2184] [General]
Hello all, 

I would like to use PyBaMM to model the electrochemical and thermal behavior of a cell during a simple experiment. Later on, I would like to compute the sensitivities of some outputs with respect to some input parameters in order to do some optimization. However, I am struggling implementing the sensitivities calculation. My code is the following: 

```python3
import pybamm
import numpy as np
import matplotlib.pyplot as plt
from ufl import sensitivity_rhs

## Creation of the models
options = {"thermal": "lumped"}
DFN_21700 = pybamm.lithium_ion.DFN(options=options)
SPM_21700 = pybamm.lithium_ion.SPM(options=options)

## Definition of the different parameters
print(pybamm.PARAMETER_PATH)
geometry = pybamm.battery_geometry(form_factor="cylindrical")
chemistry=pybamm.parameter_sets.Mohtat2020
parameters = pybamm.ParameterValues(chemistry=chemistry)
parameters["Initial concentration in positive electrode [mol.m-3]"] = 38000.

# Parameters for the sensitivity analysis
parameters.update({'Positive particle radius [m]': "[input]"})

## Definition of the experiment
experiment = pybamm.Experiment(
    ["Discharge at 1C for 2 minutes (1 second period)",
    "Rest for 20 minutes (1 second period)"]*10
)

## Discretization of the model
var_pts = {"x_n": 30, "x_s": 30, "x_p": 30, "r_n": 10, "r_p": 10}

## Running the models and comparing
pybamm.set_logging_level("NOTICE")
      
# Solve the DFN model
sim_DFN = pybamm.Simulation(DFN_21700, experiment=experiment, parameter_values=parameters, var_pts=var_pts)
sim_DFN.solve(initial_soc=1, inputs={'Positive particle radius [m]': 5E-06}, calculate_sensitivities=True)

# Solve the SPM model
sim_SPM = pybamm.Simulation(SPM_21700, experiment=experiment, parameter_values=parameters, var_pts=var_pts)
sim_SPM.solve(initial_soc=1, inputs={'Positive particle radius [m]': 5E-06}, calculate_sensitivities=True)

# Compare the models and plotting
solution_DFN, solution_SPM = sim_DFN.solution, sim_SPM.solution

t_DFN = solution_DFN["Time [s]"].entries
t_SPM = solution_SPM["Time [s]"].entries

# Temperature sensitivities w.r.t. particle size cathode
dT_dr_DFN = solution_DFN["Cell temperature [K]"].sensitivities['Positive particle radius [m]']
dT_dr_SPM = solution_SPM["Cell temperature [K]"].sensitivity['Positive particle radius [m]']
```

It seems that the "calculate_sensitivities" is not a valid argument but I am not sure what would be the appropriate syntax. 

Any help or advice would be greatly appreciated! Please let me know if you need any additional information for what I am trying to do. 

Thank you so much!

P.S.: it also seems that argument "initial_soc" is not considered and I am not sure why.

Alex

**Comment by brosaplanella** [comment]
Hi! What PyBaMM version are you using and what error do you get?

There is an example on how to calculate sensitivities in [this notebook](https://github.com/pybamm-team/workshop-notebooks/blob/main/2021-09-workshop/Session%203/sensitivities_and_data_fitting.ipynb)

**Reply by alexgUCSD** [reply]
Hi, 

Thank you for your answer!
I am using PyBaMM version 22.6 which I think is the latest. I am getting the following error: 
`  File "power_profile.py", line 112, in <module>
    sim_DFN.solve(calc_esoh=False, inputs={'Positive particle radius [m]': 5E-06}, calculate_sensitivities=True)
  File "/home/alex/.local/lib/python3.8/site-packages/pybamm/simulation.py", line 815, in solve
    step_solution = solver.step(
TypeError: step() got an unexpected keyword argument 'calculate_sensitivities'`

I have seen the notebook and this is what I am trying to reproduce but I have been unsuccessful so far. 

Thanks,
Alex

**Reply by valentinsulzer** [reply]
You can't calculate sensitivities when using an experiment. Tbh I'm not sure it's possible when there are events (e.g. voltage cutoff) - definitely not possible with the way it's implemented right now

**Reply by alexgUCSD** [reply]
I see, thank you for the answer! So I should modify the example by removing the experiment and have `sim.solve([0, 3600], initial_soc=1, inputs={'Positive particle radius [m]': 5E-06}, calculate_sensitivities=True)` for instance?

**Reply by valentinsulzer** [reply]
Yes
